parquet-variant/src/main/java/org/apache/parquet/variant/VariantConverters.java

```diff
@@ -0,0 +1,776 @@
+  // do not allow instantiating this class
+  private VariantConverters() {}
+
+  /**
+   * Converter for a shredded Variant containing a value and/or typed_value field: either a top-level
+   * Variant column, or a nested array element or object field. The top-level converter is handled
+   * by a subclass (VariantColumnConverter) that also reads metadata.
+   *
+   * Converters for the `value` and `typed_value` fields are implemented as nested classes in this
+   * class.
+   *
+   * All converters for a Variant column append their results to a VariantBuilder as values are read from Parquet.
+   *
+   * Values in `typed_value` are appended by the child converter. Values in `value` are stored by a
+   * child converter, but only appended when completing this group. Additionally, object fields are
+   * appended by the `typed_value` converter, but because residual values are stored in `value`, this
+   * converter is responsible for finalizing the object.
+   */
+  static class VariantElementConverter extends GroupConverter implements VariantConverter {
+
+    private int valueIdx = -1;
+    private int typedValueIdx = -1;
+    protected Converter[] converters;
+    // The binary produces by the `value` field of this converter.
+    private Binary variantValue = null;
+
+    // The following are only used if this is an object field.
+    private String objectFieldName = null;
+    private VariantConverter parent = null;
+
+    // True if typed_value is an object (i.e. non-list group).
+    private boolean typedValueIsObject = false;
+    // Only used if typedValueIsObject is true.
+    private Set<String> shreddedObjectKeys;
+
+    @Override
+    public VariantConverter getParent() {
+      return this.parent;
+    }
+
+    @Override
+    public VariantBuilder getBuilder() {
+      return this.parent.getBuilder();
+    }
+
+    /**
+     * Called by the `value` converter to provide a new value.
+     */
+    public void setValueBinary(Binary variantValue) {
+      this.variantValue = variantValue;
+    }
+
+    public VariantElementConverter(VariantConverter parent, GroupType variantSchema, String objectFieldName) {
+      this(parent, variantSchema);
+      this.objectFieldName = objectFieldName;
+    }
+
+    public VariantElementConverter(VariantConverter parent, GroupType variantSchema) {
+      this.parent = parent;
+      converters = new Converter[variantSchema.getFieldCount()];
+
+      List<Type> fields = variantSchema.getFields();
+
+      for (int i = 0; i < fields.size(); i++) {
+        Type field = fields.get(i);
+        String fieldName = field.getName();
+        if (fieldName.equals("value")) {
+          this.valueIdx = i;
+          if (!field.isPrimitive() || field.asPrimitiveType().getPrimitiveTypeName() != BINARY) {
+            throw new IllegalArgumentException("Value must be a binary value");
+          }
+        } else if (fieldName.equals("typed_value")) {
+          this.typedValueIdx = i;
+        }
```

Comment:
If there is no `else` case, then I think this is going to fail at read time when Parquet tries to call methods on a `null` converter. To ignore a field we would need to add a noop converter. For now, I'd recommend just failing with a good exception message, like `throw new UnsupportedOperationException("Cannot read variant with unexpected field: " + fieldName);`